src/OpenSage.Game/Logic/SelectionSystem.cs

```diff
@@ -173,20 +174,36 @@ private void MultiSelect()
 
                 if (gameObject.Collider.Intersects(boxFrustum))
                 {
-                    var objectId = (uint) Game.Scene3D.GameObjects.GetObjectId(gameObject);
-                    selectedObjectIds.Add(objectId);
+                    selectedObjects.Add(gameObject);
                 }
             }
 
+
+            var units = selectedObjects.Where(x => x.Definition.KindOf?.Get(ObjectKinds.Structure) == false).ToList();
+            var structures = selectedObjects.Where(x => x.Definition.KindOf?.Get(ObjectKinds.Structure) == true).ToList();
```

Comment:
You could do `Take(1)` here, since you only ever want the first one.

At some point we'll want to remove LINQ from game-loop code, but this is fine for now.

Reply:
I refactored the whole code, should be ready now.